pkg/driver/agent.go

```diff
@@ -68,11 +74,51 @@ func NewNode(d *CSIDriver) csi.NodeServer {
 		}
 	}()
 
+	if d.config.ListenAddress != "" {
+		exposeMetrics(d.config, stopCh)
+	}
+
 	return &node{
 		driver: d,
 	}
 }
 
+type promErrorLog struct{}
+
+func (p *promErrorLog) Println(v ...interface{}) {
+	klog.Errorln(v...)
+}
+
+func exposeMetrics(c *config.Config, stopCh <-chan struct{}) {
+	registry := prometheus.NewRegistry()
+	if err := registry.Register(collector.NewDeviceCollector(stopCh)); err != nil {
+		klog.Fatalf("failed to register device metrics collector: %v", err)
+	}
+	if !c.DisableExporterMetrics {
+		if err := registry.Register(collectors.NewProcessCollector(
+			collectors.ProcessCollectorOpts{})); err != nil {
+			klog.Fatalf("failed to register process metrics collector: %v", err)
+		}
+		if err := registry.Register(collectors.NewGoCollector()); err != nil {
+			klog.Fatalf("failed to register go runtime metrics collector: %v", err)
+		}
+	}
+
+	http.Handle(c.MetricsPath, promhttp.InstrumentMetricHandler(registry,
+		promhttp.HandlerFor(registry, promhttp.HandlerOpts{ErrorLog: &promErrorLog{}})))
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		_, _ = w.Write([]byte(`<html><head><title>Device Exporter</title></head><body>
+			<h1>Device Exporter</h1><p><a href="` + c.MetricsPath + `">Metrics</a></p>
+			</body></html>`))
+	})
+
+	go func() {
+		if err := http.ListenAndServe(c.ListenAddress, nil); err != nil && err != http.ErrServerClosed {
```

Comment:
Why do we need to check for `ErrServerClosed`?

Reply:
We should error out also for other errors right?

Reply:
muscle memory! `ErrServerClosed` is being returned if one explicitly [shuts down](https://golang.org/pkg/net/http/#Server.Shutdown) http server. But I think it is not required here since we are explicitly calling shutting down anywhere. let me remove it.